server/events/project_command_context_builder.go

```diff
@@ -283,9 +284,32 @@ func escapeArgs(args []string) []string {
 	return escaped
 }
 
+/*
+    if len(module.RequiredCore) != 1 {
+		ctx.Log.Info("cannot determine which version to use from terraform configuration, detected %d possibilities.", len(module.RequiredCore))
+		return nil
+	}
+	requiredVersionSetting := module.RequiredCore[0]
+
+	// We allow `= x.y.z`, `=x.y.z` or `x.y.z` where `x`, `y` and `z` are integers.
+	re := regexp.MustCompile(`^=?\s*([^\s]+)\s*$`)
+	matched := re.FindStringSubmatch(requiredVersionSetting)
+	if len(matched) == 0 {
+		ctx.Log.Debug("did not specify exact version in terraform configuration, found %q", requiredVersionSetting)
+		return nil
+	}
+	ctx.Log.Debug("found required_version setting of %q", requiredVersionSetting)
+	version, err := version.NewVersion(matched[1])
+	if err != nil {
+		ctx.Log.Debug(err.Error())
+		return nil
+	}
+
+*/
+
 // Extracts required_version from Terraform configuration.
 // Returns nil if unable to determine version from configuration.
-func getTfVersion(ctx *command.Context, absProjDir string) *version.Version {
+func getTfVersion(ctx *command.Context, terraformClient terraform.Client, absProjDir string) *version.Version {
```

Comment:
we may move this logic into `terraform.Client` too? This method just check directory structure and extract matched terraform version. `ctx` is used only for `ctx.Log`.

Reply:
Good point. Then it would just take in a logger and the project directory. How about:
```
DetectVersion(projectDirectory string, log logging.SimpleLogging) *version.Version 
```

Reply:
Addressed in https://github.com/runatlantis/atlantis/pull/2843/commits/2a3673833e4d4e6c32cc85e163f34b73ab29ca26

This also had the nice side-effect of splitting some testing logic. Now the terraform_client tests deal with resolving versions for various exact & non-exact contstraints, whereas the project_command_context_builder tests deal with whether the default from `atlantis.yaml` takes precedence or not.